# The Predicament at Pandas Pizzeria

You've been hired on as an assistant for Pandas Pizzeria, the premiere pizza place in the village of Analyston. You enter the establishment and are immediately greeted by dishes breaking, screaming, and overall chaos. It seems that the manager and the chef are arguing over what to put on the menu, as it's a new fiscal year starting today. The manager notices you and sighs a breath of relief. "Thank goodness you're here," they say, "You can solve this for us. You said in the interview you have experience with Pandas, yes?"

In [1]:
import os
import pandas as pd

You nod, excited to finally be putting your newfound data skills to some real use.

"Excellent. I have a few files here from the last year's sales. I also have a list of questions I'd like you to answer to help us make a new menu. In general, we'll need you to join these files into one dataframe and analyze the popularity of the pizzas over different frames of time, as well as a few other helpful tidbits of information. 

Once you've done all that, if it all checks out, then I'll ask you what you think we should do specifically to improve the menu, so keep that in mind while you're doing your analysis. Sound good? Great! Good luck."

The CSV files are sent to you and you get down to business.

## Retrieving the Data

### Task 1: Load the Data from the CSV Files

Your initial task is to take the 4 csv files you have been given and turn them into pandas dataframes. A quick note - these files aren't encoded in UTF-8 so when you're reading them into dataframes be sure to use the argument `encoding = 'latin-1'`.

Hint

<details>

`.read_csv()` is a handy function.

In [2]:
orders, order_details, pizzas, pizza_types = None, None, None, None
absolute_path = os.path.abspath(os.path.dirname('archive'))

# This is for JH filesystem. It may be different on your computer.
orders = pd.read_csv(absolute_path + '/archive/orders.csv', encoding='latin-1')

### BEGIN SOLUTION
order_details = pd.read_csv(absolute_path + '/archive/order_details.csv', encoding='latin-1')
pizzas = pd.read_csv(absolute_path + '/archive/pizzas.csv', encoding='latin-1')
pizza_types = pd.read_csv(absolute_path + '/archive/pizza_types.csv', encoding='latin-1')
### END SOLUTION

In [3]:
assert orders.columns.tolist() == ['order_id', 'date', 'time']
assert order_details.columns.tolist() == ['order_details_id', 'order_id', 'pizza_id', 'quantity']
assert pizzas.columns.tolist() == ['pizza_id', 'pizza_type_id', 'size', 'price']
assert pizza_types.columns.tolist() == ['pizza_type_id', 'name', 'category', 'ingredients']

## Exploring and Shaping the Data

### Task 2: Taking a Peek at the Dataframes

Now that you have the csv files as dataframes, let's take a look at each of them. Examine the first 10 entries in each dataframe. Note: For pizza_types, since there are only about 30 entries, output the whole dataframe so you can get a look at all the types of pizza that are sold.

In [4]:
### BEGIN SOLUTION
orders.head(10)
### END SOLUTION

order_id        date      time
0         1  2015-01-01  11:38:36
1         2  2015-01-01  11:57:40
2         3  2015-01-01  12:12:28
3         4  2015-01-01  12:16:31
4         5  2015-01-01  12:21:30
5         6  2015-01-01  12:29:36
6         7  2015-01-01  12:50:37
7         8  2015-01-01  12:51:37
8         9  2015-01-01  12:52:01
9        10  2015-01-01  13:00:15

In [5]:
### BEGIN SOLUTION
order_details.head(10)
### END SOLUTION

order_details_id  order_id       pizza_id  quantity
0                 1         1     hawaiian_m         1
1                 2         2  classic_dlx_m         1
2                 3         2  five_cheese_l         1
3                 4         2    ital_supr_l         1
4                 5         2     mexicana_m         1
5                 6         2     thai_ckn_l         1
6                 7         3    ital_supr_m         1
7                 8         3   prsc_argla_l         1
8                 9         4    ital_supr_m         1
9                10         5    ital_supr_m         1

In [6]:
### BEGIN SOLUTION
pizzas.head(10)
### END SOLUTION

pizza_id pizza_type_id size  price
0      bbq_ckn_s       bbq_ckn    S  12.75
1      bbq_ckn_m       bbq_ckn    M  16.75
2      bbq_ckn_l       bbq_ckn    L  20.75
3     cali_ckn_s      cali_ckn    S  12.75
4     cali_ckn_m      cali_ckn    M  16.75
5     cali_ckn_l      cali_ckn    L  20.75
6  ckn_alfredo_s   ckn_alfredo    S  12.75
7  ckn_alfredo_m   ckn_alfredo    M  16.75
8  ckn_alfredo_l   ckn_alfredo    L  20.75
9    ckn_pesto_s     ckn_pesto    S  12.75

In [7]:
### BEGIN SOLUTION
pizza_types
### END SOLUTION

pizza_type_id                                        name category  \
0        bbq_ckn                  The Barbecue Chicken Pizza  Chicken   
1       cali_ckn                The California Chicken Pizza  Chicken   
2    ckn_alfredo                   The Chicken Alfredo Pizza  Chicken   
3      ckn_pesto                     The Chicken Pesto Pizza  Chicken   
4     southw_ckn                 The Southwest Chicken Pizza  Chicken   
5       thai_ckn                      The Thai Chicken Pizza  Chicken   
6       big_meat                          The Big Meat Pizza  Classic   
7    classic_dlx                    The Classic Deluxe Pizza  Classic   
8       hawaiian                          The Hawaiian Pizza  Classic   
9    ital_cpcllo                 The Italian Capocollo Pizza  Classic   
10    napolitana                        The Napolitana Pizza  Classic   
11   pep_msh_pep  The Pepperoni, Mushroom, and Peppers Pizza  Classic   
12     pepperoni                         The Pepperoni Pizza  Classic   
13     the_greek                             The Greek Pizza  Classic   
14    brie_carre                        The Brie Carre Pizza  Supreme   
15     calabrese                         The Calabrese Pizza  Supreme   
16     ital_supr                   The Italian Supreme Pizza  Supreme   
17  peppr_salami                     The Pepper Salami Pizza  Supreme   
18    prsc_argla            The Prosciutto and Arugula Pizza  Supreme   
19      sicilian                          The Sicilian Pizza  Supreme   
20   soppressata                       The Soppressata Pizza  Supreme   
21    spicy_ital                     The Spicy Italian Pizza  Supreme   
22  spinach_supr                   The Spinach Supreme Pizza  Supreme   
23   five_cheese                       The Five Cheese Pizza   Veggie   
24   four_cheese                       The Four Cheese Pizza   Veggie   
25  green_garden                      The Green Garden Pizza   Veggie   
26   ital_veggie                The Italian Vegetables Pizza   Veggie   
27  mediterraneo                     The Mediterranean Pizza   Veggie   
28      mexicana                          The Mexicana Pizza   Veggie   
29    spin_pesto                     The Spinach Pesto Pizza   Veggie   
30   spinach_fet                  The Spinach and Feta Pizza   Veggie   
31    veggie_veg           The Vegetables + Vegetables Pizza   Veggie   

                                          ingredients  
0   Barbecued Chicken, Red Peppers, Green Peppers,...  
1   Chicken, Artichoke, Spinach, Garlic, Jalapeno ...  
2   Chicken, Red Onions, Red Peppers, Mushrooms, A...  
3   Chicken, Tomatoes, Red Peppers, Spinach, Garli...  
4   Chicken, Tomatoes, Red Peppers, Red Onions, Ja...  
5   Chicken, Pineapple, Tomatoes, Red Peppers, Tha...  
6   Bacon, Pepperoni, Italian Sausage, Chorizo Sau...  
7   Pepperoni, Mushrooms, Red Onions, Red Peppers,...  
8            Sliced Ham, Pineapple, Mozzarella Cheese  
9   Capocollo, Red Peppers, Tomatoes, Goat Cheese,...  
10  Tomatoes, Anchovies, Green Olives, Red Onions,...  
11                Pepperoni, Mushrooms, Green Peppers  
12                       Mozzarella Cheese, Pepperoni  
13  Kalamata Olives, Feta Cheese, Tomatoes, Garlic...  
14  Brie Carre Cheese, Prosciutto, Caramelized Oni...  
15  ‘Nduja Salami, Pancetta, Tomatoes, Red Onions,...  
16  Calabrese Salami, Capocollo, Tomatoes, Red Oni...  
17  Genoa Salami, Capocollo, Pepperoni, Tomatoes, ...  
18  Prosciutto di San Daniele, Arugula, Mozzarella...  
19  Coarse Sicilian Salami, Tomatoes, Green Olives...  
20  Soppressata Salami, Fontina Cheese, Mozzarella...  
21  Capocollo, Tomatoes, Goat Cheese, Artichokes, ...  
22  Spinach, Red Onions, Pepperoni, Tomatoes, Arti...  
23  Mozzarella Cheese, Provolone Cheese, Smoked Go...  
24  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...  
25  Spinach, Mushrooms, Tomatoes, Green Olives, Fe...  
26  Eggplant, Artichokes, Tomatoes, Zucchini, Red ...  
27  Spinach, Artichokes, 

#### Question (answer in the cell below): What is the `pizza_type_id` at index 29 in `pizza_types`?

ANSWER HERE: Spin_Pesto

### Task 3: Merging the Dataframes

Looks like we have some pretty simple tables, so it should be a breeze to merge them all together for easier analysis. Let's go ahead and do that now.

Hint

<details>

`pd.merge()` will be come in handy for this one! Make sure you're merging on the right attribute for each merge.

In [11]:
merged_data = None

### BEGIN SOLUTION
merge_order1 = pd.merge(orders,order_details, on='order_id')
merge_order2 = pd.merge(merge_order1,pizzas, on='pizza_id')
merged_data = pd.merge(merge_order2,pizza_types,on='pizza_type_id')
### END SOLUTION

merged_data

order_id        date      time  order_details_id       pizza_id  \
0             1  2015-01-01  11:38:36                 1     hawaiian_m   
1             2  2015-01-01  11:57:40                 2  classic_dlx_m   
2             2  2015-01-01  11:57:40                 3  five_cheese_l   
3             2  2015-01-01  11:57:40                 4    ital_supr_l   
4             2  2015-01-01  11:57:40                 5     mexicana_m   
...         ...         ...       ...               ...            ...   
48615     21348  2015-12-31  21:23:10             48616  ckn_alfredo_m   
48616     21348  2015-12-31  21:23:10             48617  four_cheese_l   
48617     21348  2015-12-31  21:23:10             48618   napolitana_s   
48618     21349  2015-12-31  22:09:54             48619     mexicana_l   
48619     21350  2015-12-31  23:02:05             48620      bbq_ckn_s   

       quantity pizza_type_id size  price                        name  \
0             1      hawaiian    M  13.25          The Hawaiian Pizza   
1             1   classic_dlx    M  16.00    The Classic Deluxe Pizza   
2             1   five_cheese    L  18.50       The Five Cheese Pizza   
3             1     ital_supr    L  20.75   The Italian Supreme Pizza   
4             1      mexicana    M  16.00          The Mexicana Pizza   
...         ...           ...  ...    ...                         ...   
48615         1   ckn_alfredo    M  16.75   The Chicken Alfredo Pizza   
48616         1   four_cheese    L  17.95       The Four Cheese Pizza   
48617         1    napolitana    S  12.00        The Napolitana Pizza   
48618         1      mexicana    L  20.25          The Mexicana Pizza   
48619         1       bbq_ckn    S  12.75  The Barbecue Chicken Pizza   

      category                                        ingredients  
0      Classic           Sliced Ham, Pineapple, Mozzarella Cheese  
1      Classic  Pepperoni, Mushrooms, Red Onions, Red Peppers,...  
2       Veggie  Mozzarella Cheese, Provolone Cheese, Smoked Go...  
3      Supreme  Calabrese Salami, Capocollo, Tomatoes, Red Oni...  
4       Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...  
...        ...                                                ...  
48615  Chicken  Chicken, Red Onions, Red Peppers, Mushrooms, A...  
48616   Veggie  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...  
48617  Classic  Tomatoes, Anchovies, Green Olives, Red Onions,...  
48618   Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...  
48619  Chicken  Barbecued Chicken, Red Peppers, Green Peppers,...  

[48620 rows x 12 columns]

In [12]:
assert merged_data.shape == (48620, 12)
print('Success!')

Success!


### Task 4: Finding Basic Frequencies

Nice, that's a bit better! Now we have all of our data in one easily-referenced dataframe. Let's take a look at a few different columns and find the frequencies of their values. Let's look at `category` and `size`.

Hint:

<details>

Pandas Series have a method specifically to **count values**.

In [13]:
### BEGIN SOLUTION
merged_data['category'].value_counts()
### END SOLUTION

category
Classic    14579
Supreme    11777
Veggie     11449
Chicken    10815
Name: count, dtype: int64

In [14]:
### BEGIN SOLUTION
merged_data['size'].value_counts()
### END SOLUTION

size
L      18526
M      15385
S      14137
XL       544
XXL       28
Name: count, dtype: int64

### Question: From these observations, what is the most likely combination of category and size in an order?: Classic, (L) pizza because they are the highest values in there column

ANSWER HERE

### Task 5: Making New Columns

Awesome! Now we want to make a new column in our dataframe for later easy analysis - we're going to update and combine our `date` and `time` columns into a new `datetime` column.

Note: You're not deleting any columns, just updating them. Be sure you're actually updating them, too, since `pd` methods tend to be in-place operations and don't inherently change the referenced object.

Hints:

<details>

`date` is a breeze, we're just using `.to_datetime()` on it to convert it to the right format.
`time` is a little odder because we want JUST the time and when we convert it to a datetime (as you may see), it ends up including the default year of 1900 (which is inaccurate) - this is why we append .dt.time to the end of our `.to_datetime()` method - this will convert the column back into an object dtype - this is fine
`datetime` is simply a datetime-formatted combination of `date` (as a string) and `time` (as a string).


In [20]:
# Overview of tasks in this cell (each can be done in one line of code):
# 1. Convert the 'date' column to a datetime object
# 2. Convert the 'time' column to a datetime.dt.time object using `format='%H:%M:%S'` to specify the format.
#   - Include .dt.time to get the time component of the datetime object (e.g. 12:30:00)
# 3. Create a new column called 'datetime' that combines the 'date' and 'time' columns (as strings - use the `.astype()` method)

### BEGIN SOLUTION
merged_data['date'] = pd.to_datetime(merged_data['date'])
merged_data['time'] = pd.to_datetime(merged_data['time'],format='%H:%M:%S').dt.time
merged_data['datetime'] = merged_data['date'].astype(str)+ " " + merged_data['time'].astype(str)
### END SOLUTION

merged_data

order_id       date      time  order_details_id       pizza_id  \
0             1 2015-01-01  11:38:36                 1     hawaiian_m   
1             2 2015-01-01  11:57:40                 2  classic_dlx_m   
2             2 2015-01-01  11:57:40                 3  five_cheese_l   
3             2 2015-01-01  11:57:40                 4    ital_supr_l   
4             2 2015-01-01  11:57:40                 5     mexicana_m   
...         ...        ...       ...               ...            ...   
48615     21348 2015-12-31  21:23:10             48616  ckn_alfredo_m   
48616     21348 2015-12-31  21:23:10             48617  four_cheese_l   
48617     21348 2015-12-31  21:23:10             48618   napolitana_s   
48618     21349 2015-12-31  22:09:54             48619     mexicana_l   
48619     21350 2015-12-31  23:02:05             48620      bbq_ckn_s   

       quantity pizza_type_id size  price                        name  \
0             1      hawaiian    M  13.25          The Hawaiian Pizza   
1             1   classic_dlx    M  16.00    The Classic Deluxe Pizza   
2             1   five_cheese    L  18.50       The Five Cheese Pizza   
3             1     ital_supr    L  20.75   The Italian Supreme Pizza   
4             1      mexicana    M  16.00          The Mexicana Pizza   
...         ...           ...  ...    ...                         ...   
48615         1   ckn_alfredo    M  16.75   The Chicken Alfredo Pizza   
48616         1   four_cheese    L  17.95       The Four Cheese Pizza   
48617         1    napolitana    S  12.00        The Napolitana Pizza   
48618         1      mexicana    L  20.25          The Mexicana Pizza   
48619         1       bbq_ckn    S  12.75  The Barbecue Chicken Pizza   

      category                                        ingredients  \
0      Classic           Sliced Ham, Pineapple, Mozzarella Cheese   
1      Classic  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
2       Veggie  Mozzarella Cheese, Provolone Cheese, Smoked Go...   
3      Supreme  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
4       Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
...        ...                                                ...   
48615  Chicken  Chicken, Red Onions, Red Peppers, Mushrooms, A...   
48616   Veggie  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...   
48617  Classic  Tomatoes, Anchovies, Green Olives, Red Onions,...   
48618   Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
48619  Chicken  Barbecued Chicken, Red Peppers, Green Peppers,...   

                  datetime  
0      2015-01-01 11:38:36  
1      2015-01-01 11:57:40  
2      2015-01-01 11:57:40  
3      2015-01-01 11:57:40  
4      2015-01-01 11:57:40  
...                    ...  
48615  2015-12-31 21:23:10  
48616  2015-12-31 21:23:10  
48617  2015-12-31 21:23:10  
48618  2015-12-31 22:09:54  
48619  2015-12-31 23:02:05  

[48620 rows x 13 columns]

Now let's make another column called `weekday`, again for some more analysis down the line.

Hint:

<details>

- `.dt` for a Series gives us a datetime object that we can reformat.
- `.strftime()` will be your reformatting friend.

In [42]:
### BEGIN SOLUTION
merged_data['weekday'] = merged_data['date'].dt.strftime('%A')
### END SOLUTION

merged_data

order_id       date      time  order_details_id       pizza_id  \
0             1 2015-01-01  11:38:36                 1     hawaiian_m   
1             2 2015-01-01  11:57:40                 2  classic_dlx_m   
2             2 2015-01-01  11:57:40                 3  five_cheese_l   
3             2 2015-01-01  11:57:40                 4    ital_supr_l   
4             2 2015-01-01  11:57:40                 5     mexicana_m   
...         ...        ...       ...               ...            ...   
48615     21348 2015-12-31  21:23:10             48616  ckn_alfredo_m   
48616     21348 2015-12-31  21:23:10             48617  four_cheese_l   
48617     21348 2015-12-31  21:23:10             48618   napolitana_s   
48618     21349 2015-12-31  22:09:54             48619     mexicana_l   
48619     21350 2015-12-31  23:02:05             48620      bbq_ckn_s   

       quantity pizza_type_id size  price                        name  \
0             1      hawaiian    M  13.25          The Hawaiian Pizza   
1             1   classic_dlx    M  16.00    The Classic Deluxe Pizza   
2             1   five_cheese    L  18.50       The Five Cheese Pizza   
3             1     ital_supr    L  20.75   The Italian Supreme Pizza   
4             1      mexicana    M  16.00          The Mexicana Pizza   
...         ...           ...  ...    ...                         ...   
48615         1   ckn_alfredo    M  16.75   The Chicken Alfredo Pizza   
48616         1   four_cheese    L  17.95       The Four Cheese Pizza   
48617         1    napolitana    S  12.00        The Napolitana Pizza   
48618         1      mexicana    L  20.25          The Mexicana Pizza   
48619         1       bbq_ckn    S  12.75  The Barbecue Chicken Pizza   

      category                                        ingredients  \
0      Classic           Sliced Ham, Pineapple, Mozzarella Cheese   
1      Classic  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
2       Veggie  Mozzarella Cheese, Provolone Cheese, Smoked Go...   
3      Supreme  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
4       Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
...        ...                                                ...   
48615  Chicken  Chicken, Red Onions, Red Peppers, Mushrooms, A...   
48616   Veggie  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...   
48617  Classic  Tomatoes, Anchovies, Green Olives, Red Onions,...   
48618   Veggie  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
48619  Chicken  Barbecued Chicken, Red Peppers, Green Peppers,...   

                  datetime   weekday  
0      2015-01-01 11:38:36  Thursday  
1      2015-01-01 11:57:40  Thursday  
2      2015-01-01 11:57:40  Thursday  
3      2015-01-01 11:57:40  Thursday  
4      2015-01-01 11:57:40  Thursday  
...                    ...       ...  
48615  2015-12-31 21:23:10  Thursday  
48616  2015-12-31 21:23:10  Thursday  
48617  2015-12-31 21:23:10  Thursday  
48618  2015-12-31 22:09:54  Thursday  
48619  2015-12-31 23:02:05  Thursday  

[48620 rows x 14 columns]

In [30]:
assert merged_data.shape == (48620, 14)
print('Success!')

Success!


### Task 6: Checking for Types and Nulls

Great! We now have all of the columns we'll need to find some insights from this dataframe. Now let's finish off our exploratory analysis with these two checks:
- We'll check the datatype of each of the columns in our dataframe with `.info()`
- We'll check for nulls in each column with `.isnull()`

Hint:

<details>

There are a few different ways to check for null values, but if you're looking for the easiest way to see if there are any (and how many), you can just sum the result of `.isnull()` on `merged_data`.

In [32]:
### BEGIN SOLUTION
merged_data.info()
### END SOLUTION

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   order_id          48620 non-null  int64         
 1   date              48620 non-null  datetime64[ns]
 2   time              48620 non-null  object        
 3   order_details_id  48620 non-null  int64         
 4   pizza_id          48620 non-null  object        
 5   quantity          48620 non-null  int64         
 6   pizza_type_id     48620 non-null  object        
 7   size              48620 non-null  object        
 8   price             48620 non-null  float64       
 9   name              48620 non-null  object        
 10  category          48620 non-null  object        
 11  ingredients       48620 non-null  object        
 12  datetime          48620 non-null  object        
 13  weekday           48620 non-null  object        
dtypes: datetime64[ns](1), 

In [43]:
### BEGIN SOLUTION
merged_data.isnull().sum()

### END SOLUTION

order_id            0
date                0
time                0
order_details_id    0
pizza_id            0
quantity            0
pizza_type_id       0
size                0
price               0
name                0
category            0
ingredients         0
datetime            0
weekday             0
dtype: int64

#### Question - How many Dtypes do we have in this dataframe? Any nulls?

ANSWER HERE: 14 dtypes and 0 nulls

# Describing the Data

### Task 7: Generating Descriptive Statistics

We now have a nice and clean dataframe that we can use to make some predictions! Before we dive too deep into any particular facet, let's take a quick look at the overall descriptive statistics of the dataset. We'll want to use the `.describe()` method - this method defaults to analyzing only the numeric dtypes in the dataframe. This is fine for now.

In [41]:
### BEGIN SOLUTION
merged_data.describe()
### END SOLUTION

order_id                           date  order_details_id  \
count  48620.000000                          48620      48620.000000   
mean   10701.479761  2015-06-29 11:03:43.611682560      24310.500000   
min        1.000000            2015-01-01 00:00:00          1.000000   
25%     5337.000000            2015-03-31 00:00:00      12155.750000   
50%    10682.500000            2015-06-28 00:00:00      24310.500000   
75%    16100.000000            2015-09-28 00:00:00      36465.250000   
max    21350.000000            2015-12-31 00:00:00      48620.000000   
std     6180.119770                            NaN      14035.529381   

           quantity         price  
count  48620.000000  48620.000000  
mean       1.019622     16.494132  
min        1.000000      9.750000  
25%        1.000000     12.750000  
50%        1.000000     16.500000  
75%        1.000000     20.250000  
max        4.000000     35.950000  
std        0.143077      3.621789

Looking good! Now let's use the same method but add in the argument `include= "???"` - with the `"???"` being a string of a particular dtype. You should have all of the info at your disposal to do this without any further help. If you'd like more info on this method, here's the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html).

In [53]:
### BEGIN SOLUTION
merged_data.describe(include='all')
### END SOLUTION

order_id                           date      time  \
count   48620.000000                          48620     48620   
unique           NaN                            NaN     16382   
top              NaN                            NaN  12:32:00   
freq             NaN                            NaN        26   
mean    10701.479761  2015-06-29 11:03:43.611682560       NaN   
min         1.000000            2015-01-01 00:00:00       NaN   
25%      5337.000000            2015-03-31 00:00:00       NaN   
50%     10682.500000            2015-06-28 00:00:00       NaN   
75%     16100.000000            2015-09-28 00:00:00       NaN   
max     21350.000000            2015-12-31 00:00:00       NaN   
std      6180.119770                            NaN       NaN   

        order_details_id    pizza_id      quantity pizza_type_id   size  \
count       48620.000000       48620  48620.000000         48620  48620   
unique               NaN          91           NaN            32      5   
top                  NaN  big_meat_s           NaN   classic_dlx      L   
freq                 NaN        1811           NaN          2416  18526   
mean        24310.500000         NaN      1.019622           NaN    NaN   
min             1.000000         NaN      1.000000           NaN    NaN   
25%         12155.750000         NaN      1.000000           NaN    NaN   
50%         24310.500000         NaN      1.000000           NaN    NaN   
75%         36465.250000         NaN      1.000000           NaN    NaN   
max         48620.000000         NaN      4.000000           NaN    NaN   
std         14035.529381         NaN      0.143077           NaN    NaN   

               price                      name category  \
count   48620.000000                     48620    48620   
unique           NaN                        32        4   
top              NaN  The Classic Deluxe Pizza  Classic   
freq             NaN                      2416    14579   
mean       16.494132                       NaN      NaN   
min         9.750000                       NaN      NaN   
25%        12.750000                       NaN      NaN   
50%        16.500000                       NaN      NaN   
75%        20.250000                       NaN      NaN   
max        35.950000                       NaN      NaN   
std         3.621789                       NaN      NaN   

                                              ingredients  \
count                                               48620   
unique                                                 32   
top     Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
freq                                                 2416   
mean                                                  NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   
std                                                   NaN   

                   datetime weekday  
count                 48620   48620  
unique                21350       7  
top     2015-11-18 12:25:12  Friday  
freq                     21    8106  
mean                    NaN     NaN  
min                     NaN     NaN  
25%                     NaN     NaN  
50%                     NaN     NaN  
75%                     NaN     NaN  
max                     NaN     NaN  
std                     NaN     NaN

It looks like you're all set to answer the manager's list of questions! Let's get started. For each of the below tasks, print your answer and paste it into the answer cell for that particular task.

### Task 8: Orders by Weekday

The boss wants to see the total unique orders for each individual day of the week. Print out the result of this query.

BONUS (1 point): The boss would be extra happy if you ordered the weeks Monday - Friday in your response.

In [63]:
### BEGIN SOLUTION
total_orders = merged_data['weekday'].value_counts().reset_index()
total_orders.columns = ['weekday', 'total_orders']

day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
all_days = pd.DataFrame(day_order, columns=['weekday'])
new_merged_orders = all_days.merge(total_orders, on='weekday')
new_merged_orders['total_orders'] = new_merged_orders['total_orders'].astype(int)

print(new_merged_orders)
### END SOLUTION

     weekday  total_orders
0     Monday          6369
1    Tuesday          6753
2  Wednesday          6797
3   Thursday          7323
4     Friday          8106
5   Saturday          7355
6     Sunday          5917


ANSWER HERE:
     Monday          6369
    Tuesday          6753
  Wednesday          6797
   Thursday          7323
     Friday          8106
   Saturday          7355
     Sunday          5917

### Task 9: Orders by Hour

The boss would also like total unique orders by the hour. Again (and for all subsequent problems), print out your results.

In [75]:
### BEGIN SOLUTION
merged_data['hour'] = merged_data['datetime'].dt.hour
order_per_hour = merged_data.groupby('hour')['order_id'].nunique().reset_index()
order_per_hour.columns = ['hour', 'order_per_hour']
print(order_per_hour)
### END SOLUTION

    hour  order_per_hour
0      9               1
1     10               8
2     11            1231
3     12            2520
4     13            2455
5     14            1472
6     15            1468
7     16            1920
8     17            2336
9     18            2399
10    19            2009
11    20            1642
12    21            1198
13    22             663
14    23              28


ANSWER HERE:
    hour  order_per_hour
0      9               1
1     10               8
2     11            1231
3     12            2520
4     13            2455
5     14            1472
6     15            1468
7     16            1920
8     17            2336
9     18            2399
10    19            2009
11    20            1642
12    21            1198
13    22             663
14    23              28


### Task 10: Customers Per Day

The boss is curious as to what was the average number of customers each day.

In [74]:
### BEGIN SOLUTION
daily_customers = merged_data.groupby(merged_data['date'].dt.date)['order_id'].nunique().reset_index()
daily_customers.columns = ['date', 'unique_customers']
avg = daily_customers['unique_customers'].mean()
print(f"Number of orders we are expecting per day: {avg}")
### END SOLUTION

Number of orders we are expecting per day: 59.63687150837989


ANSWER HERE

### Task 11: Top and Bottom 5 Pizzas

The boss is curious about which 5 pizzas have sold the most and which 5 have sold the least.

In [78]:
### BEGIN SOLUTION
sales = merged_data.groupby('pizza_id')['quantity'].sum().reset_index()
sales.columns = ['pizza_id','total_sales']
desc_order = sales.sort_values(by='total_sales',ascending=False)

top_5 = desc_order.head(5)
low_5 = desc_order.tail(5)

print(f'Top sold 5 pizzas are:\n {top_5}')
print(f'Least sold 5 pizzas are:\n {low_5}')
### END SOLUTION

Top sold 5 pizzas are:
          pizza_id  total_sales
3      big_meat_s         1914
80     thai_ckn_l         1410
20  five_cheese_l         1409
21  four_cheese_l         1316
18  classic_dlx_m         1181
Least sold 5 pizzas are:
           pizza_id  total_sales
43      mexicana_s          162
7      calabrese_s           99
13   ckn_alfredo_s           96
23  green_garden_l           95
87   the_greek_xxl           28


ANSWER HERE:
Top sold 5 pizzas are:
          pizza_id  total_sales
3      big_meat_s         1914
80     thai_ckn_l         1410
20  five_cheese_l         1409
21  four_cheese_l         1316
18  classic_dlx_m         1181
Least sold 5 pizzas are:
           pizza_id  total_sales
43      mexicana_s          162
7      calabrese_s           99
13   ckn_alfredo_s           96
23  green_garden_l           95
87   the_greek_xxl           28

### Task 12: Total Revenue

The boss also wants a number on the total price of all pizzas sold.

In [81]:
### BEGIN SOLUTION
merged_data['total_price'] = merged_data['quantity'] * merged_data['price']
total = merged_data['total_price'].sum()
print(total)
### END SOLUTION

817860.05


ANSWER HERE:
817860.05

### Task 13: Most and Least Popular Days

What were our 5 highest and 5 lowest days in terms of pizzas sold?

In [84]:
### BEGIN SOLUTION
daily_sales = merged_data.groupby('date')['quantity'].sum().reset_index()
daily_sales.columns = ['date','total_sales']
desc_order = daily_sales.sort_values(by='total_sales',ascending=False)

top_5 = desc_order.head(5)
low_5 = desc_order.tail(5)

print(f'Top 5 days are:\n {top_5}')
print(f'Least 5 days are:\n {low_5}')
### END SOLUTION

Top 5 days are:
           date  total_sales
323 2015-11-26          266
324 2015-11-27          264
283 2015-10-15          262
184 2015-07-04          234
183 2015-07-03          213
Least 5 days are:
           date  total_sales
353 2015-12-27           89
319 2015-11-22           84
356 2015-12-30           82
355 2015-12-29           80
80  2015-03-22           77


ANSWER HERE:
Top 5 days are:
           date  total_sales
323 2015-11-26          266
324 2015-11-27          264
283 2015-10-15          262
184 2015-07-04          234
183 2015-07-03          213
Least 5 days are:
           date  total_sales
353 2015-12-27           89
319 2015-11-22           84
356 2015-12-30           82
355 2015-12-29           80
80  2015-03-22           77

### Task 14: Month by Popularity

For the final question, the boss is asking for the count of pizzas sold by month and sorted in descending order.

In [89]:
### BEGIN SOLUTION
merged_data['month'] = merged_data['date'].dt.to_period('M')
monthly_sales = merged_data.groupby('month')['quantity'].sum().reset_index()
monthly_sales.columns = ['month', 'total_sales']
desc_order = monthly_sales.sort_values(by='total_sales', ascending=False)

print(desc_order)
### END SOLUTION

      month  total_sales
6   2015-07         4392
4   2015-05         4328
10  2015-11         4266
2   2015-03         4261
0   2015-01         4232
7   2015-08         4168
3   2015-04         4151
5   2015-06         4107
1   2015-02         3961
11  2015-12         3935
8   2015-09         3890
9   2015-10         3883


ANSWER HERE:
month  total_sales
6   2015-07         4392
4   2015-05         4328
10  2015-11         4266
2   2015-03         4261
0   2015-01         4232
7   2015-08         4168
3   2015-04         4151
5   2015-06         4107
1   2015-02         3961
11  2015-12         3935
8   2015-09         3890
9   2015-10         3883

### Task 15: Analyst Recommendations (BONUS - up to 2 points)

Congrats, you've finished all the tasks set out by the manager! In the cell below feel free to add any advice you may have for Pandas Pizzeria for the coming fiscal year; Any advice that's deemed useful will be add a bonus point to your overall grade, up to 2 points.

ANSWER HERE